### PR TITLE
consul: probe consul namespace feature before using namespace api

### PR DESCRIPTION
**Pull request by shoenig**
This PR changes Nomad's wrapper around the Consul NamespaceAPI so that
it will detect if the Consul Namespaces feature is enabled before making
a request to the Namespaces API. Namespaces are not enabled in Consul OSS,
and require a suitable license to be used with Consul ENT.

Previously Nomad would check for a 404 status code when makeing a request
to the Namespaces API to "detect" if Consul OSS was being used. This does
not work for Consul ENT with Namespaces disabled, which returns a 500.

Now we avoid requesting the namespace API altogether if Consul is detected
to be the OSS sku, or if the Namespaces feature is not licensed. Since
Consul can be upgraded from OSS to ENT, or a new license applied, we cache
the value for 1 minute, refreshing on demand if expired.

Fixes https://github.com/hashicorp/nomad-enterprise/issues/575

Note that the ticket originally describes using attributes from https://github.com/hashicorp/nomad/issues/10688.
This turns out not to be possible due to a chicken-egg situation between
bootstrapping the agent and setting up the consul client. Also fun: the
Consul fingerprinter creates its own Consul client, because there is no
currently no way to pass the agent's client through the fingerprint factory.

### CODE REVIEW

File: command/agent/consul/catalog_testing.go [hunk 1]
```diff
@@ -121,7 +136,34 @@ func (c *MockAgent) Self() (map[string]map[string]interface{}, error) {
...
+	return map[string]map[string]interface{}{
+		"Config": {
+			"Datacenter": "dc1",
+			"NodeName":   "x52",
+			"NodeID":     "9e7bf42e-a0b4-61b7-24f9-66dead411f0f",
+			"Revision":   "22ce6c6ad",
+			"Server":     true,
+			"Version":    version,
+		},
+		"Stats": stats,
```
Comment (tgross) [inline command/agent/consul/catalog_testing.go:166]: This field isn't documented in the [`/v1/agent/self`](https://www.consul.io/api-docs/agent#read-configuration) docs. Do we know if that's just a docs miss or is this an unstable API field?

File: command/agent/consul/self.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+	"github.com/hashicorp/go-version"
+)
+
+// Self represents the response body from Consul /v1/agent/self API endpoint.
+// Care must always be taken to do type checks when casting, as structure could
+// potentially change over time.
+type Self = map[string]map[string]interface{}
+
+func SKU(info Self) (string, bool) {
```
Comment (tgross) [inline command/agent/consul/self.go:14]: When the `SKU` and `Namespaces` functions were in the fingerprinter, it made sense that they returned `(string, bool)`. But now that they're in a different package entirely, their signature is "weird". For `SKU` the bool is false iff the string is empty, and for `Namespaces` we're returning a stringified version of the bool.

Maybe it'd be better if `SKU` returned just a string and `Namespaces` returned just a bool, and let their callers in the fingerprint extractors and namespace client sort out returning the values the fingerprinter expects?

**Review by tgross — APPROVED** [review]
LGTM. I've left a question about the `SKU` and `Namespaces` function signatures, but not a blocker.

File: command/agent/consul/catalog_testing.go (hunk 1 above)
Comment (shoenig) [reply command/agent/consul/catalog_testing.go:166]: It's [hard-coded](https://github.com/hashicorp/consul/blob/v1.9.5/agent/agent_endpoint.go#L95) as part of the response struct - I think it's just outdated docs at this point.

File: command/agent/consul/self.go (hunk 2 above)
Comment (shoenig) [reply command/agent/consul/self.go:14]: Good idea, I'll do this

Comment (shoenig) [reply command/agent/consul/self.go:14]: Did this in 4b3ed5351

**Comment by asyd** [comment]
Hello folks, 

sorry to comment here but I have no access to nomad enterprise issues. Is that the root cause of the following terraform issue introduced by 1.1.0:

```
Error: error applying jobspec: Unexpected response code: 500 (job-submitter consul token denied: consul ACL token cannot use namespace "default")
``` 

Thanks

**Comment by shoenig** [comment]
@asyd That error message looks like [this bug](https://github.com/hashicorp/nomad/issues/10718), which should also be fixed in v1.1.1

**Pull request merged**